Adding data element fields

charliepnnl opened the pull request:
This adds three new columns to every Data Group table.
As a test, I added these data fields to the `pump_power_per_flow_rate` data element of `FluidLoop` in the `ASHRAE229.schema.yaml` file. To see the resulting markdown, go to 
https://github.com/open229/ruleset-model-report-schema/blob/adding_data_element_fields/docs229/ASHRAE229.schema.md
and search for `pump_power`.

Note:  I did NOT generate the corresponding PDF file, so the one here is the old version.

mkarpman commented:
Charlie, that looks great, except I beleive the plan was to also add the forth column - "RMR Test" to identify attributes that are targeted by the RMR tests. For greatest flexibility, this fourth column can have one of the three values - False, primary, secondary. False would mean that the attributed is not part of RMR tests, secondary would mean that it is used in test case description but not the focus of the test, and primary would mean that the attribute is the focus of the test.

mkarpman commented on schema229/markdown.py:
```diff
@@ -126,4 +146,4 @@ def write_dir(input_dir_path, output_dir_path):
   elif len(sys.argv) == 1:
-    write_dir(source_dir_path, docs_dir_path)
\ No newline at end of file
+    write_dir(source_dir_path, docs_dir_path)
```
Charlie, this looks great but we also need the fourth field "RMR Test". For the greatest flexibility it should have the following enumerations: False (attribute not used in the RMR test), Secondary (attribute is used as part of the RMR test case description but is not the target of the RMR test) and Primary (attribute is the target of RMR test). Just like the other three new fields, it will have no entiry by default which would be equivalent to False.

It is my understanding that once  this change is committeed, I will be able to create a branch off the master will these fields added and could start filling them in. Correct? Thank you!

charliepnnl replied:
@mkarpman  So just to make sure:
You want the four additional columns below in this order?

RMR Test: False, Primary, Secondary
AppG Used By TCDs: True, False 
AppG Used By TCDs: True, False 
AppG B_RMR Equals P_RMR: True, False

charliepnnl replied:
@mkarpman You can actually create a branch and start filling in these fields any time, but you won't see your fields in the markdown tables until this pull request is merged in.

JasonGlazer replied:
Is the RMR Test possibly ruleset specific? In other words should it be "AppG RMR Test"?

mkarpman replied:
I think majority/all will be generic as were were trying to avoid using any ruleset-specific elements (e.g., space conditioning category) in the App G TCDs and RDSs, but time will show. If this comes up, we could add another column to differentiate the generic tests from App G - specific tests, so I think having a single RMR Test column is good for now.

charliepnnl replied:
@mkarpman I have added RMR Test to the Data Group tables.  As before, you can view the markdown at https://github.com/open229/ruleset-model-report-schema/blob/adding_data_element_fields/docs229/ASHRAE229.schema.md

mkarpman replied:
Thanks!

charliepnnl commented:
It sounds like this is ready to be merged.
Unless I hear otherwise, I will merge this in today >= 3:00 PT.

JasonGlazer commented:
This does not produce a YAML with the new fields removed, right? Is that going to be a separate pull request?

charliepnnl commented:
@JasonGlazer Oh right.  I might as well do it here.
Any preference for the name of the stripped down YAML file?  Or would you prefer a new name for the extended YAML file?

JasonGlazer commented:
Maybe ASHRAE229.schema.yaml should be what is produced after the extra fields are stripped out and the "source" file should be called ASHRAE229_extra.schema.yaml

Does that make sense?

charliepnnl commented:
@JasonGlazer  Yes, that does make sense.  I plan to add a task to the dodo.py file to automate the process.  BTW: You might want to consider automating the PDF creation and file copying steps that you must perform manually now.  I suggest that should be a separate PR, if you do want to add that automation.

charliepnnl commented:
With the latest commits:
- The main yaml file to edit is `schema-source/ASHRAE229_extra.schema.yaml`.
- Running doit now:
  1. Strips out the extra data element fields and **temporarily** saves the resulting yaml in `schema-source/ASHRAE229.schema.yaml`
  2. Validates all the yaml schema files against the meta schema
  3. Generates the markdown files in `build/docs` associated with all the yaml files
  4. Generates json files in `build/schema` associated with all the yaml files
  5. Removes `source-schema/ASHRAE229.shcema.yaml` to avoid any confusion about which yaml file should be edited
  6. Runs pytest
- I have tried to isolate and document the places in the code that are specific to 229. Namely, I modified `dodo.py` and `markdown.py` and added a new `util_229.py` file and associated unit test file `test/test_util_229.py`.

I copied `ASHRAE229_extra.schema.md` and `ASHRAE229.shcema.md` from the `build` folder into `docs229` so you can see what they both look like. I did not copy the JSON files over nor did I generate new PDF files. The FluidLoop Data Group is the only one that I added the extra fields into, so it is the only table in ASHRAE229_extra.schema.md` that shows the extra columns. (See the note to @mkarpman below.)

@JasonGlazer 
- We could fairly easily automate the process of generating the pdfs (if you are using a python package for that) and/or copy files from `build` to `docs229` if that would be helpful to you.
- Before merging this in, I will grab your latest `source-schema/ASHRAE229.shcema.yaml` and rename it `schema-source/ASHRAE229_extra.schema.yaml`.

@mkarpman With these latest changes, the inclusion of the extra columns in the Data Group markdown tables is done on a per Data Group table basis. An extra column will be included if and only if the associated field is used at least once in that Data Group. For example, including `RMR Test: True` or `RMR Test: False` in any of the Data Elements for a Data Group is enough to have the `RMR Test` column appear in the markdown table.  (This was done to minimize the code changes required and to isolate those changes as much as possible.) I hope this is clear. I believe you were planning to always include these fields anyway, in which case you don't need to change what you are doing.

JasonGlazer commented:
@charliepnnl I merged in the lastest schema and it all seems to be working properly. Thanks for your help. 

I posted a new issue #63 for your suggestion of copying files and making PDFs, great idea!
